tests/phpunit/integration/Core/Modules/REST_Modules_ControllerTest.php

```diff
@@ -0,0 +1,521 @@
+		$this->assertEquals( 500, $response->get_status() );
+	}
+
+	public function test_info_rest_endpoint__require_valid_module_slug() {
+		$request = new WP_REST_Request( 'GET', '/' . REST_Routes::REST_ROOT . '/core/modules/data/info' );
+		$request->set_body_params(
+			array(
+				'data' => array(
+					'slug' => 'fake-module',
+				),
+			)
+		);
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 'invalid_module_slug', $response->get_data()['code'] );
+		$this->assertEquals( 500, $response->get_status() );
+	}
+
+	public function test_info_rest_endpoint__valid_module_slug() {
+		$request = new WP_REST_Request( 'GET', '/' . REST_Routes::REST_ROOT . '/core/modules/data/info' );
+		$request->set_query_params( array( 'slug' => 'analytics' ) );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertNotEmpty( $response->get_data() );
+		$this->assertEquals( 200, $response->get_status() );
+	}
+
+	public function test_notifications_rest_endpoint__no_post_method() {
+		$request  = new WP_REST_Request( 'POST', '/' . REST_Routes::REST_ROOT . '/modules/analytics/data/notifications' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 400, $response->get_status() );
+	}
+
+	public function test_notifications_rest_endpoint__require_valid_slug() {
+		$request  = new WP_REST_Request( 'POST', '/' . REST_Routes::REST_ROOT . '/modules/fake-module/data/notifications' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 404, $response->get_status() );
+	}
+
+	public function test_settings_rest_endpoint__get_method() {
+		$request  = new WP_REST_Request( 'GET', '/' . REST_Routes::REST_ROOT . '/modules/analytics/data/settings' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 200, $response->get_status() );
+	}
+
+	public function test_settings_rest_endpoint__get_invalid_slug() {
+		$request  = new WP_REST_Request( 'GET', '/' . REST_Routes::REST_ROOT . '/modules/fake-module/data/settings' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 404, $response->get_status() );
+	}
+
+	public function test_settings_rest_endpoint__post_method() {
+		$request  = new WP_REST_Request( 'POST', '/' . REST_Routes::REST_ROOT . '/modules/analytics/data/settings' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 200, $response->get_status() );
+	}
+
+	public function test_settings_rest_endpoint__post_invalid_slug() {
+		$request  = new WP_REST_Request( 'POST', '/' . REST_Routes::REST_ROOT . '/modules/fake-module/data/settings' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 404, $response->get_status() );
+	}
+
+	public function test_datapoint_rest_endpoint__get_method() {
+		$request  = new WP_REST_Request( 'GET', '/' . REST_Routes::REST_ROOT . '/modules/analytics/data/accounts-properties-profiles' );
```

Comment:
You could use `modules/fake-module/data/test-request` here instead to avoid triggering a failing request. You'll need to provide it as one of the modules to the controller's `Modules` instance (see other tests for examples) but `fake-module` is intended to be used as a valid module. A test for an invalid module slug can use something like `non-existent-module` as I think is used in a few other places.

Reply:
@aaemnnosttv So I made a few updates to the `FakeModule` so that it works with our test, mostly it's treated as a module with settings now. Let me know if that's ok.

Reply:
That's probably ok given almost all modules have settings now 👍